File: scratch/pushing_multiple_values/rxjava2/Portfolio.java
import java.util.*;
import java.util.stream.*;
import java.util.concurrent.*;
import org.json.JSONObject;

import io.reactivex.BackpressureStrategy;
import io.reactivex.Flowable;
import io.reactivex.schedulers.Schedulers;
import io.reactivex.subscribers.ResourceSubscriber;
import io.reactivex.disposables.*;

public class Portfolio {
  private final Map<String, Integer> stocks = new ConcurrentHashMap<>();
  
  public void add(String ticker, Integer qty) {
    int oldQty = 0;
    if (stocks.containsKey(ticker)) {
      oldQty = stocks.get(ticker);
    }
    stocks.put(ticker, oldQty + qty);
  }
	
  public Flowable<Double> netWorth(Flowable<JSONObject> tickers) throws Exception {
    return tickers
      .filter(tick -> stocks.containsKey(tick.getString("ticker")))
      .scan(new HashMap<String, Double>(), (acc, tick) -> {
        String ticker = tick.getString("ticker");
        acc.put(ticker, stocks.get(ticker) * tick.getDouble("price"));
        return acc;  
      })
      .map(worth -> worth.values().stream().reduce(0d, (a, e) -> a + e));
  }
  
  public static void main(String[] args) throws Exception {
    Portfolio portfolio = new Portfolio();
    portfolio.add("GOOG", 10);
    portfolio.add("AAPL", 20);
    portfolio.add("YHOO", 30);
    portfolio.add("MSFT", 40);

    RealTimeNationalStockService stockService = new RealTimeNationalStockService();
    Flowable<JSONObject> realtimePrices = stockService.asFlowable()
      .filter(json -> json.has("ticker"))
      .share();
	
    Disposable netWorth = portfolio.netWorth(realtimePrices)
      .subscribe(total -> System.out.println("NetWorth => " + total), 
        error -> System.out.println("Error => " + error),
        () -> System.out.println("DONE"));
    
    Thread.sleep(10000);
    
    netWorth.dispose();
  }
}

File: scratch/everything_is_an_event/rxjavascript/portfolio.js
const rx = require('rxjs');
const { concatMap, delay, tap, map, filter, scan, take, share } = require('rxjs/operators');

class Portfolio {
  constructor() {
    this.stocks = new Map();
  }
  add(ticker, qty) {
    const oldQty = 0;
    if (this.stocks.has(ticker)) {
      oldQty = this.stocks.get(ticker);
    }
    this.stocks.set(ticker, oldQty + qty);
  }
  netWorth(tickers) {
    return tickers.pipe(
      filter(tick => this.stocks.has(tick.ticker)),
      scan((acc, tick) => {
        const ticker = tick["ticker"];
        acc.set(ticker, this.stocks.get(ticker) * tick["price"]);
        return acc;
      }, new Map()),
      map(worth => [ ... worth.values()].reduce((a, e) => a + e), 0));
  }
}

const portfolio = new Portfolio();
portfolio.add("GOOG", 10);
portfolio.add("AAPL", 20);
portfolio.add("YHOO", 30);
portfolio.add("MSFT", 40);

const { webSocket } = require('rxjs/webSocket');

const ticker = "";
const websocketUrl = `wss://national-stock-service.herokuapp.com/stocks/realtime/${ticker}`
// const websocketUrl = `ws://localhost:5000/stocks/realtime/${ticker}`

const ws = webSocket({
  url: websocketUrl,
  WebSocketCtor: require('websocket').w3cwebsocket,
});

const tickers = ws.pipe(
  tap(message => console.info(message)),
  filter(json => json["ticker"]),
  share());

// add 2% brokerage to every price that the user sees.
const brokerage = 0.02;
tickers.pipe(
  tap(tick => console.info("Before Brokerage  => " + tick["price"])),
  map(message => {
    const brokeredPrice = message["price"] * (1 + brokerage);
    message["price"] = brokeredPrice;
    return message;
  }))
  .subscribe(tick => console.info("After Brokerage Price => " + tick["price"]),
    error => console.error("Price Error => " + error),
    () => console.info("*** Price DONE ***"));

portfolio.netWorth(tickers).subscribe(
   data => console.info("Networth =>", data),
   err => console.error("Networth Error => ", err),
   () => console.info('*** Networth DONE ****')
);

// Send to Server
const commands = [
  {
	command: function() {
	  console.info("Sending subscribe..."); 
	  ws.next({ command : "subscribe" });
  	}, 
	time: 3000
  }, 
  {
	command: function() {
	  console.info("Sending unsubscribe...");
	  ws.next({ command : "unsubscribe" });
    },
	time: 8000
  }, 
  {
	command: function() {
	  console.info("Sending complete...");
	  ws.complete(); // closes the connection
    },
	time: 2000
  },
  // {
  //	command: function() {
  // 	  console.info("Sending error...");
  // 	  ws.error({code: 4000, reason: 'App broke'}); //closes the connection
  //    },
  //    time: 2000
  // }
];

rx.from(commands)
  .pipe(concatMap(item => rx.of(item.command).pipe(delay(item.time))))
  .subscribe(cmd => cmd());